### PR TITLE
Trustx adapter

### PR DESCRIPTION
**Pull request by PWyrembak**
## Type of change

- [ ] Bugfix
- [ ] Feature
- [x] New bidder adapter   
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other

## Description of change

Adding an Trustx adapter.
Added module, test spec and integration example for Trustx Bid Adapter in hello_world.html.

- test parameters for validating bids
```
{
    bidder: 'trustx',
    params: {
        uid: 42
    }
},
{
    bidder: 'trustx',
    params: {
        uid: 43
    }
},
{
    bidder: 'trustx',
    params: {
        uid: 44
    }
},
{
    bidder: 'trustx',
    params: {
        uid: 45
    }
}
```
You can setup globalPrebidTrustxPriceType to use net price type:
```
window.globalPrebidTrustxPriceType = 'net'
```

Be sure to test the integration with your adserver using the [Hello World](/integrationExamples/gpt/hello_world.html) sample page.

- paul@trustx.org
- [x] official adapter submission

### CODE REVIEW

**Comment by PWyrembak** [comment]
Hello,
Could you please comment on the reasons of trustx-adaptor pull request being rejected?

**Comment by mkendall07** [comment]
this hasn't been reviewed yet, but the build is failing. Run `gulp test` locally to see and fix the error.

**Comment by PWyrembak** [comment]
We run gulp test and it looks like the build failing not really due to adaptor code malfunctioning:

HeadlessChrome 0.0.0 (Linux 0.0.0)  "before each" hook FAILED
	Error: Uncaught SyntaxError: Unexpected token < (http://localhost:9876/context.html:1)
HeadlessChrome 0.0.0 (Linux 0.0.0): Executed 772 of 1164 (1 FAILED) (skipped 1) (0 secs / 2.66 secs)
HeadlessChrome 0.0.0 (Linux 0.0.0)  "before each" hook FAILED
	Error: Uncaught SyntaxError: Unexpected token < (http://localhost:9876/context.html:1)
e 0.0.0 (Linux 0.0.0): Executed 772 of 1164 (1 FAILED) (skipped 1) (25.213 secs / 2.66 secs)
[13:37:40] 'test-coverage' errored after 2.08 min
[13:37:40] Error: Karma tests failed with exit code 1
    at /home/travis/build/prebid/Prebid.js/gulpfile.js:94:12
    at removeAllListeners (/home/travis/build/prebid/Prebid.js/node_modules/karma/lib/server.js:380:7)
    at Server.<anonymous> (/home/travis/build/prebid/Prebid.js/node_modules/karma/lib/server.js:391:9)
    at Server.g (events.js:291:16)
    at emitNone (events.js:91:20)
    at Server.emit (events.js:185:7)
    at emitCloseNT (net.js:1544:8)
    at _combinedTickCallback (internal/process/next_tick.js:71:11)
    at process._tickCallback (internal/process/next_tick.js:98:9)

Could you please check?

File: integrationExamples/gpt/hello_world.html [hunk 1]
```diff
@@ -22,11 +22,29 @@
 
         // Replace this object to test a new Adapter!
         bids: [{
-          bidder: 'appnexus',
+          bidder: 'trustx',
```
Comment (matthewlane) [inline integrationExamples/gpt/hello_world.html]: This page is just for testing, changes to this file aren't required in PRs so the changes here can be dropped

File: modules/trustxBidAdapter.js [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+
+    if (auids.length) {
+      var onTimeoutWorker = _makeTimeoutWorker();
+      _sendRequest(auids, placementMap, onTimeoutWorker);
+    } else if (hasBid) {
+      utils.logError(LOG_ERROR_MESS.emptyUids);
+    }
+  }
+
+  function tryToSendStat(startTime, timeout, auids, isTimeouted, error) {
```
Comment (matthewlane) [inline modules/trustxBidAdapter.js]: We don't permit bid adapters to collect information about the auction mechanics, this function should be dropped. If you'd like to collect this info please create an analytics adapter

File: modules/trustxBidAdapter.js [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+    query.push('hbstat=' + (needToSendStat ? '1' : '0'));
+
+    adloader.loadScript(reqHost + path + query.join('&'));
+  }
+
+  function _makeTimeoutWorker() {
+    var active = true;
+    var handler = function() {
+      worker.called = true;
+      $$PREBID_GLOBAL$$.offEvent('bidTimeout', handler);
```
Comment (matthewlane) [inline modules/trustxBidAdapter.js]: `$$PREBID_GLOBAL$$.offEvent` and `$$PREBID_GLOBAL$$.onEvent` are meant for publisher use and shouldn't be needed within bidder adapters

**Comment by PWyrembak** [comment]
Adding Iponweb team for visibility

PAUL WYREMBAK
TrustX<http://www.trustx.org/> | Technical Operations Manager
1350 Broadway, New York, NY 10018
+1 631.428.9992<tel:+1%20631.428.9992>
[tX Logo]<http://www.trustx.org/>

File: modules/trustxBidAdapter.js (hunk 2 above)
Comment (PWyrembak) [reply modules/trustxBidAdapter.js]: This makes sense. I would like to note that this function doesn't collect any information on the 'auction mechanics' itself- it sends TrustX header bidder information on the amount of header bids which have timeouted. Do you still think we should remove this function from the adapter code? Would an analytics adapter provide such information to the header bidder?

File: modules/trustxBidAdapter.js (hunk 3 above)
Comment (PWyrembak) [reply modules/trustxBidAdapter.js]: Could you please let us know what we should replace $$PREBID_GLOBAL$$.offEvent and $$PREBID_GLOBAL$$.onEvent with?

Comment (PWyrembak) [reply modules/trustxBidAdapter.js]: Could you please let us know what we should replace $$PREBID_GLOBAL$$.offEvent and $$PREBID_GLOBAL$$.onEvent with?

File: modules/trustxBidAdapter.js (hunk 2 above)
Comment (PWyrembak) [reply modules/trustxBidAdapter.js]: This makes sense. I would like to note that this function doesn't collect any information on the 'auction mechanics' itself- it sends TrustX header bidder information on the amount of header bids which have timeouted. Do you still think we should remove this function from the adapter code? Would an analytics adapter provide such information to the header bidder?

Comment (matthewlane) [reply modules/trustxBidAdapter.js]: Thanks for clarifying. Yes, the function should remain removed as adapters should only make the minimum amount of network calls necessary for the sake of performance. Analytic adapters do have access to the `BID_TIMEOUT` event

Comment (PWyrembak) [reply modules/trustxBidAdapter.js]: Thanks for the details. As I can see you have removed the 'trytosend' stat function. Do I understand correct that you don't need us to make any changes to the TrustX adapter code for now?

Comment (matthewlane) [reply modules/trustxBidAdapter.js]: Correct, no further changes requested from me. Next a second reviewer will be assigned per our PR process and they may have additional feedback for you before merging

File: modules/trustxBidAdapter.js [hunk 4]
```diff
@@ -0,0 +1,167 @@
+var utils = require('src/utils.js');
```
Comment (mkendall07) [inline modules/trustxBidAdapter.js]: prefer `const` or `let`

File: modules/trustxBidAdapter.js [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    } else if (!resp.bid[0]) {
+      utils.logError(LOG_ERROR_MESS.noBid);
+    }
+    return resp && resp.bid && resp.bid[0];
+  }
+
+  function _forEachPlacement(error, bid, placementCode) {
+    var bidObject;
+    if (error) {
+      bidObject = bidfactory.createBid(2);
```
Comment (mkendall07) [inline modules/trustxBidAdapter.js]: use constant value here

File: modules/trustxBidAdapter.js [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    } else {
+      bidObject = bidfactory.createBid(1);
```
Comment (mkendall07) [inline modules/trustxBidAdapter.js]: pass through the bid request object as a 2nd argument to `createBid`. See example here: https://github.com/prebid/Prebid.js/blob/master/modules/appnexusAstBidAdapter.js#L206

Comment (PWyrembak) [reply modules/trustxBidAdapter.js]: thanks for the update. do we need to apply these changes to the code by ourselves?

Comment (mkendall07) [reply modules/trustxBidAdapter.js]: Yes please make the update so we can merge your adapter. Thanks.

**Comment by PWyrembak** [comment]
Hello,
Could you please review the fixes? 
Thanks

**Comment by mkendall07** [comment]
This is merged into master. Please submit a PR to the [docs repo](https://github.com/prebid/prebid.github.io) to add a file for your adapter to the [bidders directory](https://github.com/prebid/prebid.github.io/tree/master/dev-docs/bidders) so your adapter's params will appear on the [bidders page](http://prebid.org/dev-docs/bidders.html). Thank you for contributing

**Comment by PWyrembak** [comment]
We've submitted the PR to the docs repo: https://github.com/prebid/prebid.github.io/pull/378.
Pending approval.

**Comment by PWyrembak** [comment]
Hey team,
we would like to make a simple adjustment to our TrustX adapter currently available in the repo. Can we expect you reviewing it and merging soon?

**Pull request merged**